### PR TITLE
fix: windows

### LINKED ISSUE #2078: Investigate failing windows CI job
### Component

Forge

### Have you ensured that all of these are up to date?

- [ ] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

_No response_

### What command(s) is the bug in?

_No response_

### Operating System

_No response_

### Describe the bug

> thread 'multi_runner::tests::test_cheats' panicked at 'called `Result::unwrap()` on an `Err` value: Io(SolcIoError { io: Os { code: 193, kind: Uncategorized, message: "%1 is not a valid Win32 application." }, path: "C:\\Users\\runneradmin\\.svm\\0.6.12\\solc-0.6.12" })', forge\src\test_helpers.rs:45:36

https://github.com/foundry-rs/foundry/runs/7006059778?check_suite_focus=true

perhaps this is some solc win32 snafu, perhaps with the 0.6.12 version

from googling "%1 is not a valid Win32 application.", this could be an issue with the PATH var or a general windows OS related issue https://stackoverflow.com/questions/185042/how-do-i-resolve-1-is-not-a-valid-win32-application

### LINKED ISSUE #2085: forge script error
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (024e3ed 2022-06-23T03:30:48.7987658Z)

### What command(s) is the bug in?

forge script

### Operating System

Windows

### Describe the bug

Facing an issue when it comes to using forge script. Replicated the solidity scripting tutorial in the foundry book for testing purposes and am still facing the same error

Running `forge script script/NFT.s.sol:MyScript --fork-url http://localhost:8545 --private-key $PRIVATE_KEY_ANVIL` outputs 

```
[⠊] Compiling...
[⠆] Compiling 1 files with 0.8.14
[⠰] Solc 0.8.14 finished in 8.85s
Compiler run successful
The application panicked (crashed).
Message:  Target not found?
Location: cli\src\cmd\forge\script\build.rs:163

This is a bug. Consider reporting it at https://github.com/foundry-rs/foundry

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
```

Able to bypass this error by omitting the script name by running `forge script script/NFT.s.sol --fork-url http://localhost:8545 --private-key $PRIVATE_KEY_ANVIL` instead. 

### Steps to Reproduce Error
1. Clone my sample repo [here](https://github.com/wuchuank/foundry-script)
2. Run anvil and add PRIVATE_KEY_ANVIL as an environment variable, alternatively just copy and paste from anvil into the command
3. Run `forge script script/NFT.s.sol:MyScript --fork-url http://localhost:8545 --private-key $PRIVATE_KEY_ANVIL`

### LINKED ISSUE #2092: forge test --debug crashes: "Target not found?"
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (7b9d3ad 2022-06-17T21:57:41.4016897Z)

### What command(s) is the bug in?

forge test

### Operating System

Windows

### Describe the bug

I'm trying to debug a test for a contract using `forge test` but i can't do so for some reason, `forge debug` runs just fine:

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

contract Contract {
    uint256 public count = 0;

    function plus() public {
        count += 1;
    }
}
```

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

import "forge-std/Test.sol";
import "src/Contract.sol";


contract ContractTest is Test {
    Contract c;
    
    function setUp() public {
        c = new Contract();
    }

    function testCount() public {
        assertEq(c.count(), 0);
        c.plus();
        assertEq(c.count(), 1);
    }
}
```

Trying to run: `forge test --debug "testCount"` gives me this:

```
C:\Users\proto\Desktop\foundry-play>forge test --debug testCount
[⠔] Compiling...
[⠊] Compiling 1 files with 0.8.15
[⠒] Solc 0.8.15 finished in 1.49s
Compiler run successful
[⠔] Compiling...
Nothing to compile
The application panicked (crashed).
Message:  Target not found?
Location: cli\src\cmd\forge\script\build.rs:163

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
```

while using `forge debug --debug test/Contract.t.sol --sig "testCount()"` works just fine, not sure if i am missing something here?

### PR DESCRIPTION
**Pull request by onbjerg**
After the svm-rs fix we now have new failing tests. Also adds a new test for #2092 and #2085, and will (eventually) fix those bugs

Closes #2078, closes #2092, closes #2085

### CODE REVIEW

**Comment by onbjerg** [comment]
Upon more investigation it seems the tests fail because we try to match run tests using `ArtifactId::identifier` as keys, and on Windows these use \, whereas on other platforms they use /

**Comment by mattsse** [comment]
> using ArtifactId::identifier as keys,

Where exactly? Where do we perform comparisons with "identifiers"?

**Comment by onbjerg** [comment]
@mattsse In `MultiContractRunner` we key the results sent back to the test reporter by `ArtifactId::identifier` here: https://github.com/foundry-rs/foundry/blob/9dcc0952b8f2e98c84e357205af45e9def535bb9/forge/src/multi_runner.rs#L287

In the multirunner tests we collect these results in a `BTreeMap` and try to compare them with what we expect. The expected results are stored in a `BTreeMap`, again, keyed by `ArtifactId::identifier`: https://github.com/foundry-rs/foundry/blob/9dcc0952b8f2e98c84e357205af45e9def535bb9/forge/src/multi_runner.rs#L488

**Comment by mattsse** [comment]
I see, but this is only an issue with the test, right? because on windows, everything should be \\?

we can update the tests with either `format!("src{}..", MAIN_SEPARATOR)`

See https://doc.rust-lang.org/std/path/constant.MAIN_SEPARATOR.html

I think it's also worthwhile adding a function to ArtifactId that does convert slashes, via path-slash crate

**Comment by onbjerg** [comment]
Currently failing things:

- ANSI escape sequences in our snapshots, but those are not present on Windows
- Cast crashes because of a stack overflow for some reason (Forge doesn't lol?)
- Slashes in remappings are wrong
- Slashes in some snapshots are wrong (from `ArtifactId::identifier`)

File: cli/src/cmd/forge/script/build.rs [hunk 1]
```diff
@@ -99,7 +99,7 @@ impl ScriptArgs {
 
         let mut extra_info = ExtraLinkingInfo {
             no_target_name,
-            target_fname,
+            target_fname: target_fname.clone(),
```
Comment (onbjerg) [inline cli/src/cmd/forge/script/build.rs:102]: This clone is for a better error message further down.. Once we clean up linking we should consider cleaning up the targetting logic in script as well (and other commands that take `<path>:<name>`)

File: cli/src/cmd/forge/script/build.rs [hunk 2]
```diff
@@ -148,9 +145,11 @@ impl ScriptArgs {
                         extra.target_id = Some(id.clone());
                     }
                 } else {
-                    let split: Vec<&str> = extra.target_fname.split(':').collect();
-                    let path = std::path::Path::new(split[0]);
-                    let name = split[1];
+                    let (path, name) = extra
+                        .target_fname
+                        .rsplit_once(':')
+                        .expect("The target specifier is malformed.");
```
Comment (onbjerg) [inline cli/src/cmd/forge/script/build.rs:151]: Turns out the bug was staring us in the face all along - on Windows paths obviously contain `:` so the path we compared with in this branch was literally just the drive letter on Windows lmao

File: cli/src/cmd/forge/script/build.rs [hunk 3]
```diff
@@ -210,22 +211,25 @@ impl ScriptArgs {
 
         // We received `contract_path:contract_name`
         if let Some(path) = contract.path {
-            let path = dunce::canonicalize(&path)?;
+            let path =
+                dunce::canonicalize(&path).wrap_err("Could not canonicalize the target path")?;
```
Comment (onbjerg) [inline cli/src/cmd/forge/script/build.rs:215]: Some error message improvements

File: cli/src/utils.rs [hunk 4]
```diff
@@ -246,10 +246,12 @@ pub(crate) use p_println;
 /// Disables terminal colours if either:
 /// - Running windows and the terminal does not support colour codes.
 /// - Colour has been disabled by some environment variable.
+/// - We are running inside a test
 pub fn enable_paint() {
+    let is_test = cfg!(test);
     let is_windows = cfg!(windows) && !Paint::enable_windows_ascii();
     let env_colour_disabled = std::env::var("NO_COLOR").is_ok();
-    if is_windows || env_colour_disabled {
+    if is_test || is_windows || env_colour_disabled {
```
Comment (onbjerg) [inline cli/src/utils.rs:254]: Windows doesn't support color out of the box so instead of replacing every ANSI escape code in our fixtures I just opted to disable colors in tests. I think this is OK, if the colors are off it's not critical and we'd know easily anyway

File: cli/test-utils/src/util.rs [hunk 5]
```diff
@@ -635,16 +635,16 @@ pub trait OutputExt {
...
-    Regex::new(r"(finished in (.*)?s|-->(.*).sol|Location(.|\n)*\.rs(.|\n)*Backtrace|installing solc version(.*?)\n|Successfully installed solc(.*?)\n)").unwrap()
+    Regex::new(r"(\r|finished in (.*)?s|-->(.*).sol|Location(.|\n)*\.rs(.|\n)*Backtrace|installing solc version(.*?)\n|Successfully installed solc(.*?)\n)").unwrap()
 });
 
 impl OutputExt for process::Output {
     #[track_caller]
     fn stdout_matches_path(&self, expected_path: impl AsRef<Path>) -> &Self {
         let expected = fs::read_to_string(expected_path).unwrap();
-        let expected = IGNORE_IN_FIXTURES.replace_all(&expected, "");
+        let expected = IGNORE_IN_FIXTURES.replace_all(&expected, "").replace('\\', "/");
```
Comment (onbjerg) [inline cli/test-utils/src/util.rs:645]: This is because `ArtifactId::identifier` does not contain a canonicalized path

File: cli/test-utils/src/util.rs [hunk 6]
```diff
@@ -689,6 +689,21 @@ pub fn dir_list<P: AsRef<Path>>(dir: P) -> Vec<String> {
...
+/// be canonicalized.
+pub fn remapping_str(src: &str, dest: &str) -> String {
+    // NOTE(onbjerg): The `trim_end_matches` is because the path itself on Windows is normalized
+    // except for the last character which is still a /...
+    format!(
+        "{}={}/",
+        src,
+        dest.trim_end_matches('/')
+            .replace('/', std::str::from_utf8(&[std::path::MAIN_SEPARATOR as u8]).unwrap())
+    )
```
Comment (onbjerg) [inline cli/test-utils/src/util.rs:704]: Remappings from `Config` on Windows seem to use Windows paths, *except* if part of the path is from another remapping (see the config tests for examples).

The remappings we use to compare in tests however are not canonicalized. I'm not sure what the best fix is given how flexible remappings are

File: cli/tests/it/cmd.rs [hunk 7]
```diff
@@ -426,8 +433,24 @@ warning[5667]: Warning: Unused function parameter. Remove or comment out the var
     ));
 });
 
-// tests that direct import paths are handled correctly
-forgetest!(canhandle_direct_imports_into_src, |prj: TestProject, mut cmd: TestCommand| {
+// Tests that direct import paths are handled correctly
+//
+// NOTE(onbjerg): Disabled for Windows -- for some reason solc fails with a bogus error message
```
Comment (onbjerg) [inline cli/tests/it/cmd.rs:438]: I really have NO IDEA at all what is going on here. "Invalid implicit conversion from struct Bar memory to struct Bar memory"????

File: cli/tests/it/config.rs [hunk 8]
```diff
@@ -420,12 +422,25 @@ forgetest_init!(can_detect_lib_foundry_toml, |prj: TestProject, mut cmd: TestCom
...
         vec![
-            "another-lib/=lib/nested-lib/lib/another-lib/src/".parse().unwrap(),
-            "ds-test/=lib/forge-std/lib/ds-test/src/".parse().unwrap(),
-            "forge-std/=lib/forge-std/src/".parse().unwrap(),
-            "nested-lib/=lib/nested-lib/src/".parse().unwrap(),
-            "nested-twice/=lib/nested-lib/lib/another-lib/lib/nested-twice/".parse().unwrap(),
-            "nested/=lib/nested-lib/lib/nested/".parse().unwrap(),
+            // NOTE(onbjerg): For some reason a part of this remapping was not normalized on
+            // Windows
+            format!("{}another-lib/src/", remapping_str("another-lib/", "lib/nested-lib/lib/"))
```
Comment (onbjerg) [inline cli/tests/it/config.rs:427]: Not sure what's going on here... I think it might be because the path contains other remappings? It's a bit hard to follow

File: cli/tests/it/config.rs [hunk 9]
```diff
@@ -436,12 +451,28 @@ forgetest_init!(can_detect_lib_foundry_toml, |prj: TestProject, mut cmd: TestCom
...
         vec![
-            "another-lib/=lib/nested-lib/lib/another-lib/custom-source-dir/".parse().unwrap(),
-            "ds-test/=lib/forge-std/lib/ds-test/src/".parse().unwrap(),
-            "forge-std/=lib/forge-std/src/".parse().unwrap(),
-            "nested-lib/=lib/nested-lib/src/".parse().unwrap(),
-            "nested-twice/=lib/nested-lib/lib/another-lib/lib/nested-twice/".parse().unwrap(),
-            "nested/=lib/nested-lib/lib/nested/".parse().unwrap(),
+            // NOTE(onbjerg): For some reason a part of this remapping was not normalized on
+            // Windows
+            format!(
```
Comment (onbjerg) [inline cli/tests/it/config.rs:456]: Same as above

File: fmt/src/formatter.rs [hunk 10]
```diff
@@ -2775,9 +2775,15 @@ mod tests {
         source: &str,
         expected_source: &str,
     ) {
-        #[derive(PartialEq, Eq)]
+        #[derive(Eq)]
         struct PrettyString(String);
 
+        impl PartialEq for PrettyString {
+            fn eq(&self, other: &PrettyString) -> bool {
+                self.0.lines().eq(other.0.lines())
```
Comment (onbjerg) [inline fmt/src/formatter.rs:2783]: Avoids having to replace `\r\n` with `\n`

**Comment by onbjerg** [comment]
Finally! I reverted the workflow back to its original state(ish), but a successful run can be found here: https://github.com/foundry-rs/foundry/runs/7165826844

File: cli/test-utils/src/script.rs [hunk 11]
```diff
@@ -88,17 +88,17 @@ impl ScriptTester {
         }
     }
 
-    fn link_testdata(current_dir: &Path) -> eyre::Result<()> {
+    fn copy_testdata(current_dir: &Path) -> eyre::Result<()> {
         let testdata = format!("{}/../../testdata", env!("CARGO_MANIFEST_DIR"));
-        std::fs::hard_link(
+        std::fs::copy(
```
Comment (onbjerg) [inline cli/test-utils/src/script.rs:93]: Hard links from some non-temp directory to a temp directory is considered a violation of filesystem boundaries on Windows, so we have to copy instead 🙄

File: cli/src/cmd/forge/script/build.rs (hunk 2 above)
Comment (mattsse) [reply cli/src/cmd/forge/script/build.rs:151]: omg

**Review by mattsse — CHANGES_REQUESTED** [review]
awesome!

not sure what to do about the `remapping_str`.
perhaps easiest solution would be using `path_slash::PathBufExt` to convert all slashes in the remappings?

File: cli/test-utils/src/util.rs [hunk 12]
```diff
@@ -689,6 +692,21 @@ pub fn dir_list<P: AsRef<Path>>(dir: P) -> Vec<String> {
         .collect()
 }
 
+/// Creates a cross-platform remapping string for use in tests
+///
+/// NOTE: This probably should be unnecessary, and remappings should probably
+/// be canonicalized.
+pub fn remapping_str(src: &str, dest: &str) -> String {
```
[the remark this answers is not in the document]
Comment (onbjerg) [reply cli/test-utils/src/util.rs:699]: I wonder if we should do that in ethers-solc instead - not sure if using / would break anything internally, but I think at a minimum it would be safe to use / in `Display`?

**Comment by gakonst** [comment]
High level lgtm. Wonder if we should think about it more or ship it and let Windows users tell us what works well and what doesn't?

**Comment by onbjerg** [comment]
@gakonst I think ship and figure it out, there are probably other bugs lurking but it's hard to find

File: cli/test-utils/src/util.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (mattsse) [reply cli/test-utils/src/util.rs:699]: that makes sense!
